task_processing/plugins/mesos/logging_executor.py

```diff
@@ -0,0 +1,151 @@
+import json
+import logging
+import sys
+import time
+import urllib.request
+from threading import Thread
+
+from pyrsistent import field
+from pyrsistent import m
+from pyrsistent import PMap
+from pyrsistent import pmap
+from pyrsistent import PRecord
+from six.moves.queue import Queue
+
+from task_processing.interfaces.task_executor import TaskExecutor
+
+log = logging.getLogger(__name__)
+
+# Read task log in 4K chunks
+TASK_LOG_CHUNK_LEN = 4096
+
+
+class LogMetadata(PRecord):
+    log_url = field(type=str, initial='')
+    log_offsets = field(type=PMap,
+                        factory=pmap,
+                        initial=pmap({'stdout': 0, 'stderr': 0}))
+
+
+class LoggingExecutor(TaskExecutor):
+    def __init__(self, downstream_executor):
+        self.downstream_executor = downstream_executor
+
+        self.src_queue = downstream_executor.get_event_queue()
+        self.dest_queue = Queue()
+        self.stopping = False
+
+        # No lock is needed because log_metadata is touched in a single thread
+        self.log_metadata = m()
+
+        self.logging_thread = Thread(target=self.logging_loop)
+        self.logging_thread.daemon = True
+        self.logging_thread.start()
+
+    def build_task_log_url(self, task_log_url, event):
+        try:
+            response = json.load(urllib.request.urlopen(
+                task_log_url + 'files/debug'))
+        except Exception as e:
+            log.error("Failed to fetch files {error}".format(error=e))
+            return ''
+
+        for key in response.keys():
+            if event.raw.executor_id.value in key and \
+                    event.raw.container_status.container_id.value in key:
+                return (task_log_url + 'files/read?path=' + key + '/')
+        return ''
+
+    def stream_task_log(self, task_id):
+        log_md = self.log_metadata[task_id]
+        # Do not fetch logs if url hasn't been built
+        if log_md.log_url is '':
+            return
+
+        offsets = {
+            'stdout': log_md.log_offsets['stdout'],
+            'stderr': log_md.log_offsets['stderr']
+        }
+        for f in ['stdout', 'stderr']:
+            offset = offsets[f]
+            while True:
+                url = log_md.log_url + f + \
+                    '&length=' + str(TASK_LOG_CHUNK_LEN) + \
+                    '&offset=' + str(offset)
+                try:
+                    response = json.load(urllib.request.urlopen(url))
+                    log_length = len(response['data'])
+                    for line in response['data'].splitlines():
+                        print(
+                            task_id + ": " + line,
```

Comment:
I think we're going to want the log line to be as 'rich' as possible.

Can we do things like format it as

iso8601datetime task_id[container_id@agent_ip]: logline

Reply:
Will fix except for the iso time. Applications should add timestamps to their logs instead. The executor timestamp is not accurate.